From 29f68fd09e3202732073424f5173d2a09fe64a73 Mon Sep 17 00:00:00 2001
From: Michal Simek <monstr@monstr.eu>
Date: Fri, 11 Feb 2011 13:24:46 +0100
Subject: [PATCH] microblaze: Fix init ramfs section alignment

.init.ramfs.info section wasn't aligned properly.

Kernel log:
<0>Kernel panic - not syncing: no cpio magic

readelf dump below shows the issue.

$ microblaze-unknown-linux-gnu-readelf -S wrong
There are 47 section headers, starting at offset 0x24e7c0c:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        c0000000 001000 1d2d80 00  AX  0   0 16
  [ 2] __fdt_blob        PROGBITS        c01d2d80 1d3d80 004000 00   A  0   0  1
  [ 3] .rodata           PROGBITS        c01d7000 1d8000 0af597 00   A  0   0 32
  [ 4] .init.ramfs.info  PROGBITS        c0286597 287597 000004 00   A  0   0  1
  [ 5] __jump_table      PROGBITS        c02865a0 c2f204 000000 00   W  0   0  1
  [ 6] .pci_fixup        PROGBITS        c02865a0 c2f204 000000 00   W  0   0  1

Signed-off-by: Michal Simek <monstr@monstr.eu>
---
 arch/microblaze/kernel/vmlinux.lds.S |   16 +++-------------
 1 files changed, 3 insertions(+), 13 deletions(-)

diff --git a/arch/microblaze/kernel/vmlinux.lds.S b/arch/microblaze/kernel/vmlinux.lds.S
index 96a88c3..3451bde 100644
--- a/arch/microblaze/kernel/vmlinux.lds.S
+++ b/arch/microblaze/kernel/vmlinux.lds.S
@@ -123,20 +123,10 @@ SECTIONS {
 
 	__init_end_before_initramfs = .;
 
-	.init.ramfs ALIGN(PAGE_SIZE) : AT(ADDR(.init.ramfs) - LOAD_OFFSET) {
-		__initramfs_start = .;
-		*(.init.ramfs)
-		__initramfs_end = .;
-		. = ALIGN(4);
-		LONG(0);
-/*
- * FIXME this can break initramfs for MMU.
- * Pad init.ramfs up to page boundary,
- * so that __init_end == __bss_start. This will make image.elf
- * consistent with the image.bin
- */
-		/* . = ALIGN(PAGE_SIZE); */
+	.init.ramfs : AT(ADDR(.init.ramfs) - LOAD_OFFSET) {
+		INIT_RAM_FS
 	}
+
 	__init_end = .;
 
 	.bss ALIGN (PAGE_SIZE) : AT(ADDR(.bss) - LOAD_OFFSET) {
-- 
1.7.0.4

